kernel/aster-nix/src/net/socket/vsock/stream/init.rs

```diff
@@ -0,0 +1,111 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use crate::{
+    events::{IoEvents, Observer},
+    net::socket::vsock::{
+        addr::{VsockSocketAddr, VMADDR_CID_ANY, VMADDR_PORT_ANY},
+        VSOCK_GLOBAL,
+    },
+    prelude::*,
+    process::signal::{Pollee, Poller},
+};
+
+pub struct Init {
+    bind_addr: SpinLock<Option<VsockSocketAddr>>,
+    pollee: Pollee,
+}
+
+impl Init {
+    pub fn new() -> Self {
+        Self {
+            bind_addr: SpinLock::new(None),
+            pollee: Pollee::new(IoEvents::empty()),
+        }
+    }
+
+    pub fn bind(&self, addr: VsockSocketAddr) -> Result<()> {
+        if self.bind_addr.lock().is_some() {
+            return_errno_with_message!(Errno::EINVAL, "the socket is already bound");
+        }
+        let vsockspace = VSOCK_GLOBAL.get().unwrap();
+
+        // check correctness of cid
+        let local_cid = vsockspace.driver.lock_irq_disabled().guest_cid();
+        if addr.cid != VMADDR_CID_ANY && addr.cid != local_cid as u32 {
+            return_errno_with_message!(Errno::EADDRNOTAVAIL, "The cid in address is incorrect");
+        }
+        let mut new_addr = addr;
+        new_addr.cid = local_cid as u32;
+
+        // check and assign a port
+        if addr.port == VMADDR_PORT_ANY {
+            if let Ok(port) = vsockspace.alloc_ephemeral_port() {
+                new_addr.port = port;
+            } else {
+                return_errno_with_message!(Errno::EAGAIN, "cannot find unused high port");
+            }
+        } else if vsockspace
+            .used_ports
+            .lock_irq_disabled()
+            .contains(&new_addr.port)
+        {
+            return_errno_with_message!(Errno::EADDRNOTAVAIL, "the port in address is occupied");
+        } else {
+            vsockspace
+                .used_ports
+                .lock_irq_disabled()
+                .insert(new_addr.port);
+        }
```

Comment:
This usage will cause race condition.

You check the used port in line 47-50 and then drop the lock. And then re-acquire lock in line 54 to insert port. But if someone modified the `used_port` after dropping the lock and before re-acquiring the lock, the check result is invalid.

So the lock should be held all the time whenchecking and inserting.